dynamic.Message cannot unmarshal type map<string, string>

I have this proto message 

```
message GetRequest {
  message Data {
    map<string, string> headers = 1;
    string data = 2;
  }
  Data data = 1;
}
```

And it causing panic when I'm supplying the header value. Is this expected?

```
panic: interface conversion: map[string]string is not protoiface.MessageV1: missing method ProtoMessage

goroutine 8 [running]:
github.com/jhump/protoreflect/dynamic.(*KnownTypeRegistry).CreateIfKnown(0x1400063af60, {0x14000716200, 0x31})
	/Users/sahal/Workspace/go/pkg/mod/github.com/jhump/protoreflect@v1.12.0/dynamic/message_factory.go:158 +0x168
github.com/jhump/protoreflect/dynamic.(*MessageFactory).NewMessage(0x1400048a670, 0x14000620c30)
	/Users/sahal/Workspace/go/pkg/mod/github.com/jhump/protoreflect@v1.12.0/dynamic/message_factory.go:64 +0x4c
github.com/jhump/protoreflect/codec.DecodeLengthDelimitedField(0x14000598b80, {0x14000636334, 0x20, 0x24}, {0x100fba5c0, 0x1400048a670})
	/Users/sahal/Workspace/go/pkg/mod/github.com/jhump/protoreflect@v1.12.0/codec/decode_fields.go:195 +0xc0
github.com/jhump/protoreflect/codec.(*Buffer).decodeKnownField(0x1400036b0c8, 0x14000598b80, 0x2, {0x100fba5c0, 0x1400048a670})
	/Users/sahal/Workspace/go/pkg/mod/github.com/jhump/protoreflect@v1.12.0/codec/decode_fields.go:270 +0x160
github.com/jhump/protoreflect/codec.(*Buffer).DecodeFieldValue(0x1400036b0c8, 0x1400036b000, {0x100fba5c0, 0x1400048a670})
	/Users/sahal/Workspace/go/pkg/mod/github.com/jhump/protoreflect@v1.12.0/codec/decode_fields.go:109 +0xac
github.com/jhump/protoreflect/dynamic.(*Message).unmarshal(0x1400063b230, 0x1400036b0c8, 0x0)
	/Users/sahal/Workspace/go/pkg/mod/github.com/jhump/protoreflect@v1.12.0/dynamic/binary.go:154 +0x74
github.com/jhump/protoreflect/dynamic.(*Message).UnmarshalMerge(...)
	/Users/sahal/Workspace/go/pkg/mod/github.com/jhump/protoreflect@v1.12.0/dynamic/binary.go:149
github.com/jhump/protoreflect/dynamic.(*Message).Unmarshal(0x1400063b230, {0x14000636332, 0x26, 0x26})
	/Users/sahal/Workspace/go/pkg/mod/github.com/jhump/protoreflect@v1.12.0/dynamic/binary.go:138 +0x6c
google.golang.org/protobuf/internal/impl.legacyUnmarshal({{}, {0x100fd7fd8, 0x1400048a730}, {0x14000636332, 0x26, 0x26}, 0x0, {0x100fc53a0, 0x14000194e40}})
	/Users/sahal/Workspace/go/pkg/mod/google.golang.org/protobuf@v1.26.0/internal/impl/legacy_message.go:419 +0x150
google.golang.org/protobuf/proto.UnmarshalOptions.unmarshal({{}, 0x1, 0x1, 0x0, {0x100fc53a0, 0x14000194e40}}, {0x14000636332, 0x26, 0x26}, {0x100fd7fd8, ...})
	/Users/sahal/Workspace/go/pkg/mod/google.golang.org/protobuf@v1.26.0/proto/decode.go:93 +0x13c
google.golang.org/protobuf/proto.UnmarshalOptions.UnmarshalState(...)
	/Users/sahal/Workspace/go/pkg/mod/google.golang.org/protobuf@v1.26.0/proto/decode.go:66
github.com/golang/protobuf/proto.UnmarshalMerge({0x14000636332, 0x26, 0x26}, {0x100fc6130, 0x1400063b230})
	/Users/sahal/Workspace/go/pkg/mod/github.com/golang/protobuf@v1.5.2/proto/wire.go:67 +0xd8
github.com/golang/protobuf/proto.Unmarshal({0x14000636332, 0x26, 0x26}, {0x100fc6130, 0x1400063b230})
	/Users/sahal/Workspace/go/pkg/mod/github.com/golang/protobuf@v1.5.2/proto/wire.go:58 +0x5c
github.com/jhump/protoreflect/codec.DecodeLengthDelimitedField(0x14000598b00, {0x14000636332, 0x26, 0x26}, {0x100fba5c0, 0x1400048a670})
	/Users/sahal/Workspace/go/pkg/mod/github.com/jhump/protoreflect@v1.12.0/codec/decode_fields.go:196 +0xf8
github.com/jhump/protoreflect/codec.(*Buffer).decodeKnownField(0x1400036b688, 0x14000598b00, 0x2, {0x100fba5c0, 0x1400048a670})
	/Users/sahal/Workspace/go/pkg/mod/github.com/jhump/protoreflect@v1.12.0/codec/decode_fields.go:270 +0x160
github.com/jhump/protoreflect/codec.(*Buffer).DecodeFieldValue(0x1400036b688, 0x1400036b5c0, {0x100fba5c0, 0x1400048a670})
	/Users/sahal/Workspace/go/pkg/mod/github.com/jhump/protoreflect@v1.12.0/codec/decode_fields.go:109 +0xac
github.com/jhump/protoreflect/dynamic.(*Message).unmarshal(0x1400063af90, 0x1400036b688, 0x0)
	/Users/sahal/Workspace/go/pkg/mod/github.com/jhump/protoreflect@v1.12.0/dynamic/binary.go:154 +0x74
github.com/jhump/protoreflect/dynamic.(*Message).UnmarshalMerge(...)
	/Users/sahal/Workspace/go/pkg/mod/github.com/jhump/protoreflect@v1.12.0/dynamic/binary.go:149
github.com/jhump/protoreflect/dynamic.(*Message).Unmarshal(0x1400063af90, {0x14000636330, 0x28, 0x28})
	/Users/sahal/Workspace/go/pkg/mod/github.com/jhump/protoreflect@v1.12.0/dynamic/binary.go:138 +0x6c
google.golang.org/protobuf/internal/impl.legacyUnmarshal({{}, {0x100fd7fd8, 0x1400048a690}, {0x14000636330, 0x28, 0x28}, 0x0, {0x100fc53a0, 0x14000194e40}})
	/Users/sahal/Workspace/go/pkg/mod/google.golang.org/protobuf@v1.26.0/internal/impl/legacy_message.go:419 +0x150
google.golang.org/protobuf/proto.UnmarshalOptions.unmarshal({{}, 0x1, 0x1, 0x0, {0x100fc53a0, 0x14000194e40}}, {0x14000636330, 0x28, 0x28}, {0x100fd7fd8, ...})
	/Users/sahal/Workspace/go/pkg/mod/google.golang.org/protobuf@v1.26.0/proto/decode.go:93 +0x13c
google.golang.org/protobuf/proto.UnmarshalOptions.UnmarshalState(...)
	/Users/sahal/Workspace/go/pkg/mod/google.golang.org/protobuf@v1.26.0/proto/decode.go:66
github.com/golang/protobuf/proto.UnmarshalMerge({0x14000636330, 0x28, 0x28}, {0x100fc6130, 0x1400063af90})
	/Users/sahal/Workspace/go/pkg/mod/github.com/golang/protobuf@v1.5.2/proto/wire.go:67 +0xd8
github.com/golang/protobuf/proto.Unmarshal({0x14000636330, 0x28, 0x28}, {0x100fc6130, 0x1400063af90})
	/Users/sahal/Workspace/go/pkg/mod/github.com/golang/protobuf@v1.5.2/proto/wire.go:58 +0x5c
google.golang.org/grpc/encoding/proto.codec.Unmarshal({}, {0x14000636330, 0x28, 0x28}, {0x100fb2900, 0x1400063af90})
	/Users/sahal/Workspace/go/pkg/mod/google.golang.org/grpc@v1.45.0/encoding/proto/proto.go:53 +0x60
google.golang.org/grpc.(*Server).processUnaryRPC.func2({0x100fb2900, 0x1400063af90})
	/Users/sahal/Workspace/go/pkg/mod/google.golang.org/grpc@v1.45.0/server.go:1259 +0x1f0
github.com/monlabs/grpc-mock/pkg/server/mock.(*Server).createUnaryServerHandler.func1({0x0, 0x0}, {0x100fc9340, 0x1400063af00}, 0x1400010ecc0, 0x0)
	/Users/sahal/Workspace/go/src/github.com/monlabs/grpc-mock/pkg/server/mock/server.go:95 +0x128
google.golang.org/grpc.(*Server).processUnaryRPC(0x14000018700, {0x100fd3d48, 0x14000001860}, 0x140004fc000, 0x1400063a9f0, 0x1400063a690, 0x0)
	/Users/sahal/Workspace/go/pkg/mod/google.golang.org/grpc@v1.45.0/server.go:1282 +0xc5c
google.golang.org/grpc.(*Server).handleStream(0x14000018700, {0x100fd3d48, 0x14000001860}, 0x140004fc000, 0x0)
	/Users/sahal/Workspace/go/pkg/mod/google.golang.org/grpc@v1.45.0/server.go:1619 +0xa34
google.golang.org/grpc.(*Server).serveStreams.func1.2(0x1400060e100, 0x14000018700, {0x100fd3d48, 0x14000001860}, 0x140004fc000)
	/Users/sahal/Workspace/go/pkg/mod/google.golang.org/grpc@v1.45.0/server.go:921 +0x94
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/Users/sahal/Workspace/go/pkg/mod/google.golang.org/grpc@v1.45.0/server.go:919 +0x1f0
```

--- a/dynamic/message_factory.go
+++ b/dynamic/message_factory.go
@@ -174,28 +174,33 @@ func isWellKnownType(t reflect.Type) bool {
 // GetKnownType will return the reflect.Type for the given message name if it is
 // known. If it is not known, nil is returned.
 func (r *KnownTypeRegistry) GetKnownType(messageName string) reflect.Type {
-	var msgType reflect.Type
 	if r == nil {
 		// a nil registry behaves the same as zero value instance: only know of well-known types
 		t := proto.MessageType(messageName)
 		if t != nil && isWellKnownType(t) {
-			msgType = t
+			return t
 		}
-	} else {
-		if r.includeDefault {
-			msgType = proto.MessageType(messageName)
-		} else if !r.excludeWkt {
-			t := proto.MessageType(messageName)
-			if t != nil && isWellKnownType(t) {
-				msgType = t
-			}
+		return nil
+	}
+
+	if r.includeDefault {
+		t := proto.MessageType(messageName)
+		if t != nil && isMessage(t) {
+			return t
 		}
-		if msgType == nil {
-			r.mu.RLock()
-			msgType = r.types[messageName]
-			r.mu.RUnlock()
+	} else if !r.excludeWkt {
+		t := proto.MessageType(messageName)
+		if t != nil && isWellKnownType(t) {
+			return t
 		}
 	}
 
-	return msgType
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+	return r.types[messageName]
+}
+
+func isMessage(t reflect.Type) bool {
+	_, ok := reflect.Zero(t).Interface().(proto.Message)
+	return ok
 }

--- a/dynamic/message_factory_test.go
+++ b/dynamic/message_factory_test.go
@@ -66,6 +66,12 @@ func TestKnownTypeRegistry_WithDefaults(t *testing.T) {
 	checkKnownTypes(t, ktr, (*descriptor.DescriptorProto)(nil), (*descriptor.FileDescriptorProto)(nil), (*testprotos.TestMessage)(nil))
 }
 
+func TestKnownTypeRegistry_WithDefaults_MapEntry(t *testing.T) {
+	ktr := NewKnownTypeRegistryWithDefaults()
+	msgType := ktr.GetKnownType("testprotos.MapKeyFields.SEntry")
+	testutil.Require(t, msgType == nil, "should not be a known type for map entry but got %v", msgType)
+}
+
 func checkKnownTypes(t *testing.T, ktr *KnownTypeRegistry, knownTypes ...proto.Message) {
 	for _, kt := range knownTypes {
 		md, err := desc.LoadMessageDescriptorForMessage(kt)

